### ISSUE TITLE
RFC: Improvements to performance and code approachability

### ISSUE THREAD

**Issue by angeloashmore** [prismicio/prismic-gatsby#514]
Prismic’s Gatsby integration has generally been stable since its last major release, v5.0, in October 2021. As we look toward the future, however, there is still much to improve upon within the integration’s plugins.

The next major releases of `gatsby-source-prismic` and `gatsby-plugin-prismic-previews` will take a [Mac OS X Snow Leopard-like approach](https://zerodistraction.com/the-last-stable-release-of-os-x-was-snow-leopard-mountain-lion-changes-this/): no new features (well… almost none), instead focusing on performance and refactoring their codebases.

If you want to jump ahead to see plans for the next major versions, **scroll down to the Plans section**.

If you want to understand the reasoning behind the decision to focus on performance and refactoring, read on.

# Current Issues

TL;DR:

- Both plugins’ code is too complex for external contributors to read and contribute to.
- Both plugins’ performance, especially `gatsby-plugin-prismic-preview`, should be better.

## Unnecessarily complex codebase

Both `gatsby-source-prismic` and `gatsby-plugin-prismic-previews` are written in a style that makes it difficult to understand at a glance. It makes heavy use of functional programming styles using [`fp-ts`](https://gcanti.github.io/fp-ts/) which, if you aren’t familiar with its concepts, looks and reads very different from typical JavaScript.

As a result, the code is likely slower than it should be and, more importantly, does not welcome external contributors. Diagnosing or fixing a bug becomes the responsibility of the plugins’ maintainer(s), often with a delay in implementation.

If the codebase was more readable and idiomatic, bugs might be easier to resolve and new features could be implemented more quickly, potentially by developers using the plugins.

### Start-up time is slower than it needs to be

Gatsby project start-up time is significantly slower compared to other frameworks like Next.js or tools like Vite. The Prismic source plugin only impacts a portion of Gatsby’s start-up time, but it currently do not perform any optimizations to reduce its impact. Every time a Gatsby project is started or built, it must query and process every document. The time it takes to do that adds up quickly, especially for large repositories.

### Initial bundle size is too large

Supporting client-side previews means shipping preview-specific code to the client. Today, all of that code is eagerly sent to the client, even when a preview is inactive. This behavior leads to slow time to interactive sites, which is reflected in tools like Lighthouse.

Note that the Prismic Toolbar cannot be omitted as it serves an important role in managing the Prismic preview cookie that tracks preview states. Bundle size improvements will come from code directly in the preview plugin; the Prismic Toolbar will not be changing. Toolbar improvements may come at a future time, but they are outside the scope of this RFC.

### Poor performance when previewing content from large repositories (> 500 documents)

The preview plugin currently performs the following logic at the start of a preview:

1. Fetches all documents from a Prismic repository, including draft content from the preview session.
2. Transforms each document into something that appears like data from `gatsby-source-prismic`.
3. Stores all documents in a React context provider.
4. Replaces Prismic documents in page queries with data from the context provider.

When previewing content from small repositories (< 500 documents), this flow works quickly enough. Previews are loaded within a few seconds for these repositories.

Large repositories (> 500 documents), however, experience much longer preview load times due to steps 1 and 2. Querying hundreds or thousands of documents at the start of each preview and draft update can take many seconds or minutes, depending on the size of the repository. Similarly, transforming that many documents can be computationally intensive.

The current preview algorithm is inefficient, both in content fetching and preparation. The decisions made leading to this algorithm were driven by limitations in Prismic’s API. Changes to Prismic’s API are out of scope for this RFC, so any solution will need to work around the same limitations present during the plugins’ initial development.

### Configuration can be confusing

Configuring the preview plugin, notably when using the higher order components like `withPrismicUnpublishedPreview()`, can be confusing. Providing an object of page components to `withPrismicUnpublishedPreview()`, for example, is confusing, requiring an obtuse `componentResolverFromMap()` helper function.

Options were designed for complete developer flexibility, but resulted in complex options that make them difficult to use. In most cases, options are not necessary. Instead, internal behavior can rely on good defaults or assumptions.

# Plan

**Both plugins will be rewritten**. They will use idiomatic TypeScript code, notably without `fp-ts`. The rewrite has two focuses:

1. Make the codebase more accessible to external contributors.
2. Improve performance.

The source and preview plugins will continue to be supported as first-party Prismic integrations. The codebases will be reoriented toward community contributions to support timely bug fixes and improvements. This reorientation does not change Prismic’s level of support for the plugin. Rather, it is intended to provide the opportunity for external contributors to submit contributions as they see fit, something that is currently difficult.

Planned changes for each plugin are detailed below.

## `gatsby-source-prismic`

The source plugin provides content from a Prismic repository to a Gatsby website. It includes support for Gatsby Cloud’s Incremental Builds and Preview features. For websites using `gatsby-plugin-prismic-previews`, the source plugin pre-computes data at build-time needed to support client-side previews.

The rewrite will address the source plugin’s external contributor and performance issues with the following changes:

- Cache as many network requests as possible to reduce start up time. If content does not change between development server starts or production builds, for example, all content can be loaded from a local cache.

- Simplify communication between `gatsby-source-prismic` and `gatsby-plugin-prismic-previews`. The source plugin pre-computes data and makes it available in an intermediate state, requiring close coordination between the plugins. The preview plugin may be able to derive most of its data itself with less dependence from the source plugin.

- Explore a more integrated Gatsby Cloud previewing experience. Using Gatsby Cloud Previews currently requires publishing content to a temporary Prismic Release before previewing. A previewing experience that more closely resembles the experience of `gatsby-plugin-prismic-previews`, where one only needs to click the Preview button within Prismic, may be possible. Implementation of this task may be out of scope of this RFC depending on exploration results.

## `gatsby-plugin-prismic-previews`

The preview plugin provides a way to support Prismic previews in a static Gatsby site *without the need for a running server*. This means hosting a Gatsby site using simple services like Amazon S3, Netlify, and Cloudflare can support previews without the need for Gatsby Cloud.

The rewrite will address the plugin’s external contributor and performance issues with the following changes:

- Implement the existing behavior with less code and without `fp-ts`.

- Reduce the number of API calls needed to resolve a preview. There may be clever ways to use a build-time timestamp in combination with a build-time snapshot of all documents. Note that Prismic’s API does not implement a changeset-like API, so it is not possible to simply ask the API for a list of changes since build-time.

- Reduce the number of runtime document data transformations as an effect of the previous optimization. The data transformations themselves might be optimizable as well via parallelization and lazy evaluation.

- Benchmark preview performance with large repositories (at least 5000 documents). Preview performance of large repositories should be comparable to that of small repositories.

- Minimize use of custom UI elements and styling to reduce the bundle size. Custom UI elements are displayed when previews are loading, requesting an access token from the user, or displaying an error. Some of these interactions cannot be removed, but the code required to support them can be optimized.

- Integrate Prismic Toolbar’s client-side events to refresh draft updates more quickly. A full page refresh should not be necessary.

# Status of this project

Work on the above plan has already started. It is still in a *pre-*pre-release development stage and was performed primarily to validate the plan.

Although work has already begun, it is not too late to provide feedback.

# How to provide feedback

This RFC presents a high-level strategy for the next major versions of `gatsby-source-prismic` and `gatsby-plugin-prismic-previews`. It does not present implementation details. Comments about either, however, are welcomed.

If you would like to provide feedback, please reply to this issue. Everything mentioned in this RFC is open for discussion.

Thanks! 🙂

**Comment by lihbr** [comment]
LGTM, just one thing as I'm not sure it's encapsulated by the RFC yet (maybe it was implied):

> Cache as many network requests as possible to reduce start-up time.

Could we consider caching the queried & processed API locally based on the API ref (one step further than just Network cache)? Maybe this was already implied.

Going one step further, we could also explore extending cache based on the last publication date of each document when the ref has changed if the processing step is significant. However, I can see edge cases with this approach already (content relationships especially), would help in the production/preview context though(?)

**Comment by angeloashmore** [comment]
> Could we consider caching the queried & processed API locally based on the API ref (one step further than just Network cache)? Maybe this was already implied.
>
> Going one step further, we could also explore extending cache based on the last publication date of each document when the ref has changed if the processing step is significant. However, I can see edge cases with this approach already (content relationships especially), would help in the production/preview context though(?)

Yes, exactly. The most straightforward approach is to cache based on the `ref`. The `@prismicio/client` `fetch` function looks like this currently:

```ts
async (input, init) => {
	const resolvedFetch = options.fetch || fetch;
	const url = new URL(input);

	if (/\/documents\/search\/?/.test(url.pathname)) {
		return await cachedFetch(input, init, {
			fetch: resolvedFetch,
			cache: args.cache,
			name: "sourceNodes",
		});
	} else {
		return await resolvedFetch(input, init);
	}
},
```

Where `cachedFetch()` returns cached results using the URL as the key. Since the URL contains the `ref`, it automatically cache busts. The only initial `/api/v2` request is necessary to fetch the latest ref.

**Site note**: this strategy could be used elsewhere outside Gatsby projects. Imagine starting up a Next.js or Nuxt dev server and navigating around the site without making live network calls.

> Going one step further, we could also explore extending cache based on the last publication date of each document when the ref has changed if the processing step is significant. However, I can see edge cases with this approach already (content relationships especially), would help in the production/preview context though(?)

Unfortunately, we won't be able to use `last_publication_date`-based caching since it won't take into account deleted documents. We would need a proper changelog-style API (e.g. "add this document, "delete this document") to perform timestamp caching.

Perhaps a secondary service combined with Prismic's webhooks could be used to generate a third-party repository changelog. Considering this integration is primary designed for static Gatsby websites, requiring a live service would not be ideal.

**Comment by kb1995** [comment]
Hey @angeloashmore, I just wanted to see what's the progress with this RFC. The preview plugin's bundle size is damaging my Core Vital score so would be great to see if there are any upgrades in the near future. 

Also, not sure if relevant to this RFC, but I remember that last year around this time, you mentioned that Prismic has been working on Slice Machine integration which would be launched in Q1 2022. Is Prismic still planning of launching this integration or is the official Prismic recommendation to use Next.js if they want to use Slice Machine?

**Comment by angeloashmore** [comment]
Hey @kb1995, the RFC has been implemented in a private GitHub repository. We gave early access to a group of Prismic + Gatsby developers through Prismic's Early Access Program ([details here](https://prismic.io/blog/eap-gatsby-refresh-2022)). I sent you an invitation to the repository so you can try it out as well.

We plan to publish the new plugins before the end of the year. Documentation and example projects will need to be updated before publishing.

The preview plugin has been rearchitected to lazy-load everything it can. If you aren't in an active preview session, you won't load any preview-specific code except for the code that checks if you're in a preview. We also have recommendations on how you can lazy-load your page templates and Slices. Note that the Prismic toolbar will always need to be loaded.

---

> Also, not sure if relevant to this RFC, but I remember that last year around this time, you mentioned that Prismic has been working on Slice Machine integration which would be launched in Q1 2022. Is Prismic still planning of launching this integration or is the official Prismic recommendation to use Next.js if they want to use Slice Machine?

I mentioned we were working on a Slice Machine integration and hoped to release it, but unfortunately it didn't happen (details in [this PR](https://github.com/prismicio/slice-machine/pull/146) and [this issue](https://github.com/prismicio/slice-machine/issues/331)).

We're currently working on making Slice Machine pluggable, meaning we will be able to support frameworks other than Next.js and Nuxt. I have been pushing to get Gatsby support in Slice Machine for a while now (as you know), and I believe we can add support soon after we have plugin support.

I won't make any promises, but know that Gatsby with Slice Machine is still on the table.

For now, Slice Machine can officially be used only with Next.js and Nuxt. However, you can unofficially use it with Gatsby with some features disabled. [See this guide for details](https://prismic.notion.site/How-to-use-Slice-Machine-with-Gatsby-today-cbcec696392b42ad86cad6916749f6d5).

**Comment by cometkim** [comment]
> no new features (well… almost none), instead focusing on performance and refactoring their codebases.

Does this mean that there are no breaking changes in v6.0.0-alpha that require migration work? In that case, I can proceed with the upgrade right away and give some feedback.

**Comment by grahamd711** [comment]
@angeloashmore do you think you are still on track to release updated Gatsby Source Prismic by end of year? we have a large repository so the current Preview is taking considerable time to generate each time.

very are excited to hear about Slice Machine + Gatsby. appreciate your support on these great plugins, thank you!

**Comment by grahamd711** [comment]
Hi @angeloashmore hope all is well. I know there have been other internal priorities on the Prismic side, but just wanted to check in on these improvements.

The initial bundle size is killing our site speed and Lighthouse scores when we upgrade to Gatsby 4 or 5. Because of this, we are still stuck on Gatsby 3 with a fair amount of custom code splitting to reduce load times.

Is there a more defined timeline for making the performance updates? this way, we can weigh the pros and cons on our own. appreciate any insight you have!

**Comment by grahamd711** [comment]
hi @angeloashmore I saw you linked a PR to this issue last week. Would this resolve performance issues? Both gatsby-source-prismic and gatsby-plugin-prismic-previews will be re written?
